### PR TITLE
Add the quorum intersection checker code from stellar-core

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
It also includes tests.

Requires:
- ~#662~
- ~#663~

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
The part that assigns unassigned nodes needs to be removed. Not all nodes need to be in other nodes' quorums.

**Comment by Geod24** [comment]
I rebased it on v0.x.x, let's see if it passes on Windows

**Comment by Geod24** [comment]
Looks nice (and green). Will give it a review today.

File: source/scpd/quorum/QuorumTracker.d [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        quorum_tracker_rebuild(cast(void*)this, &call, cast(void*)&lookup);
+    }
+
+    // returns the current known quorum
+    ref const(QuorumMap) getQuorum() const;
+}
+
+private void quorum_tracker_rebuild (void* quorum_tracker, void* cb, void* dg);
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumTracker.d]: Use `ref` and the proper type instead of `void*`

File: source/scpd/quorum/QuorumTracker.d [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        static extern(C++) SCPQuorumSetPtr call (void* func, const ref NodeID node_id)
+        {
+            auto dg = cast(SCPQuorumSetPtr delegate(const ref NodeID)*)func;
+            return (*dg)(node_id);
+        }
+
+        quorum_tracker_rebuild(cast(void*)this, &call, cast(void*)&lookup);
+    }
+
+    // returns the current known quorum
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumTracker.d]:
```suggestion
    // returns the quorum map this quorum tracker was built with
```

File: source/scpd/quorum/QuorumTracker.d [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    // and add its dependencies as defined by `qset`
+    // returns true if expansion succeeded
+    //     `id` was unknown
+    //     `id` was known and didn't have a quorumset
+    // returns false on failure
+    // if expand fails, the caller should instead use `rebuild`
+    bool expand(const ref NodeID id, SCPQuorumSetPtr qSet);
+
+    // replaces 'rebuild' taking an std::function so we can use it with D delegates
+    extern (D) public final void rebuild (SCPQuorumSetPtr delegate(const ref NodeID) lookup)
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumTracker.d]:
```suggestion
    extern (D) public final void rebuild (SCPQuorumSetPtr delegate(const ref NodeID) lookup) @trusted
```

And other attributes.

File: source/scpd/quorum/QuorumTracker.d [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+module scpd.quorum.QuorumTracker;
+
+import scpd.Cpp;
+import scpd.scp.SCP;
+import scpd.types.Stellar_SCP;
+import scpd.types.Stellar_types;
+
+extern (C++, `stellar`):
+
+// helper class to help track the overall quorum over time
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumTracker.d]: We could improve this documentation. It was written by Stellar devs, but it's hard to parse.

Edit: do it upstream

File: source/scpp/src/quorum/QuorumTracker.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+#include "quorum/QuorumTracker.h"
+#include "scp/LocalNode.h"
+
+namespace stellar
+{
+
+// workaround to be able to pass D delegates to C++
+void quorum_tracker_rebuild (void* qt, void* cb, void* dg)
+{
+    typedef SCPQuorumSetPtr (*Callback)(void*, NodeID const&);
```
Comment (AndrejMitrovic) [inline source/scpp/src/quorum/QuorumTracker.cpp]: This function body has too much whitespace.

Comment (AndrejMitrovic) [reply source/scpp/src/quorum/QuorumTracker.cpp]: Also as mentioned before, use `ref` and proper types, not pointers.

**Review by AndrejMitrovic — COMMENTED** [review]
Completed self-review.

File: source/scpd/quorum/QuorumIntersectionChecker.d [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+extern (C++, `stellar`):
+
+extern(C++, class) public abstract class QuorumIntersectionChecker
+{
+  public:
+    extern(D) static shared_ptr!QuorumIntersectionChecker create (
+        QuorumTracker.QuorumMap map) nothrow pure @trusted @nogc
+    {
+        return create(*cast(void**)&map);
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumIntersectionChecker.d]: Should be changed to take a reference.

File: source/scpd/quorum/QuorumIntersectionChecker.d [hunk 7]
```diff
@@ -0,0 +1,36 @@
+/*******************************************************************************
+
+    Bindings for quorum/QuorumIntersectionChecker.h
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumIntersectionChecker.d]: 2020

File: source/scpd/quorum/QuorumIntersectionChecker.d [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import scpd.Cpp;
+import scpd.types.Stellar_types;
+import scpd.quorum.QuorumTracker;
+
+extern (C++, `stellar`):
+
+extern(C++, class) public abstract class QuorumIntersectionChecker
+{
+  public:
+    extern(D) static shared_ptr!QuorumIntersectionChecker create (
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumIntersectionChecker.d]: Since this is a method we introduced, we should document it.

File: source/scpd/quorum/QuorumTracker.d [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    // replaces 'rebuild' taking an std::function so we can use it with D delegates
+    extern (D) public final void rebuild (SCPQuorumSetPtr delegate(const ref NodeID) lookup)
+    {
+        static extern(C++) SCPQuorumSetPtr call (void* func, const ref NodeID node_id)
+        {
+            auto dg = cast(SCPQuorumSetPtr delegate(const ref NodeID)*)func;
+            return (*dg)(node_id);
+        }
+
+        quorum_tracker_rebuild(cast(void*)this, &call, cast(void*)&lookup);
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumTracker.d]: Replace pointers with references.

File: source/scpd/quorum/QuorumTracker.d [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    // attempts to expand quorum at node `id`
+    // expansion here means adding `id` to the known quorum
+    // and add its dependencies as defined by `qset`
+    // returns true if expansion succeeded
+    //     `id` was unknown
+    //     `id` was known and didn't have a quorumset
+    // returns false on failure
+    // if expand fails, the caller should instead use `rebuild`
+    bool expand(const ref NodeID id, SCPQuorumSetPtr qSet);
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumTracker.d]: Use our styling `expand (`

File: source/scpp/README.md [hunk 11]
```diff
@@ -15,6 +15,8 @@ Some files have been modified after porting:
 
 - Some `private` declarations were changed to `public`. This allows us to test size & layout checks for our D glue layer, for example: https://github.com/bpfkorea/agora/blob/76e4ca7d77b123ef867d3ce99c1c1b26afab761d/source/scpp/src/crypto/ByteSlice.h#L21 used in https://github.com/bpfkorea/agora/blob/76e4ca7d77b123ef867d3ce99c1c1b26afab761d/source/scpp/extra/DSizeChecks.cpp and https://github.com/bpfkorea/agora/blob/76e4ca7d77b123ef867d3ce99c1c1b26afab761d/source/scpp/extra/DLayoutChecks.cpp.
 
+- The `src/quorum` folder adapts the Quorum intersection tool which is used in Stellar's Herder implementation (not SCP). The `update.d` script does not track these files yet, so they have to be kept manually in sync when updating the bindings.
```
Comment (AndrejMitrovic) [inline source/scpp/README.md]:
```suggestion
- The `src/quorum` folder contains the Quorum intersection tool which is used in Stellar's Herder implementation (not SCP). The `update.d` script does not track these files yet, so they have to be kept manually in sync when updating the bindings.
```

**Comment by AndrejMitrovic** [comment]
I suggest adding a in-between commit which just checks in all the intersection checker code as-is, but without bindings. And change the commit author to be the Stellar devs. Otherwise it appears like the code was forged or stolen, and we don't want to leave that kind of impression on the community.

File: source/scpd/quorum/QuorumIntersectionTests.d [hunk 12]
```diff
@@ -0,0 +1,744 @@
...
+import scpd.types.XDRBase;
+
+import agora.common.Config;
+import agora.common.crypto.Key;
+
+const xvector!SCPQuorumSet nullvec;
+const shared_ptr!SCPQuorumSet nullquorum;
+
+// we replace the use of std::pair with a fixed-length array
+size_t first (size_t[2] pair) { return pair[0]; }
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumIntersectionTests.d]: This is one example where the code was changed from upstream. So we should definitely have a separate commit including all the C++ code, and then the conversion step commits after that.

File: source/scpd/quorum/QuorumIntersectionTests.d [hunk 13]
```diff
@@ -0,0 +1,744 @@
...
+    return makeSharedSCPQuorumSet(scp);
+}
+
+uint roundUpPct (size_t n, size_t pct)
+{
+    return cast(uint)(size_t(1) +
+        (((n * pct) - size_t(1)) / size_t(100)));
+}
+
+NodeID toStellarKey (PublicKey address)
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumIntersectionTests.d]: This can be replaced soon.

File: source/scpd/quorum/QuorumIntersectionTests.d [hunk 14]
```diff
@@ -0,0 +1,744 @@
...
+
+    qm[pkA.toStellarKey] = makeSharedQuorumSet(2, [pkB, pkC, pkD], null);
+    qm[pkB.toStellarKey] = makeSharedQuorumSet(2, [pkA, pkC, pkD], null);
+    qm[pkC.toStellarKey] = makeSharedQuorumSet(2, [pkA, pkB, pkD], null);
+    qm[pkD.toStellarKey] = makeSharedQuorumSet(2, [pkA, pkB, pkC], null);
+
+    auto qic = QuorumIntersectionChecker.create(qm);
+    assert(qic.networkEnjoysQuorumIntersection());
+
+    //CircularAppender().printConsole();
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumIntersectionTests.d]: Dead code. And the new test-suite logging fixes already handle this case.

File: source/scpd/quorum/QuorumIntersectionTests.d [hunk 15]
```diff
@@ -0,0 +1,744 @@
...
+}
+
+//TEST_CASE("quorum intersection 6-node with subquorums",
+//          "[herder][quorumintersection]")
```
Comment (AndrejMitrovic) [inline source/scpd/quorum/QuorumIntersectionTests.d]: This stuff in the brackets can all be removed for all of these tests.

**Comment by AndrejMitrovic** [comment]
I've split this up into multiple commits. The stellar devs get the proper attribution, and the diff is much easier to understand now.

**Comment by AndrejMitrovic** [comment]
I will fix up the rest of the comments.

**Comment by AndrejMitrovic** [comment]
Resolved all comments.

**Comment by Geod24** [comment]
> Otherwise it appears like the code was forged or stolen, and we don't want to leave that kind of impression on the community.

Just for reference: We have [this README](https://github.com/bpfkorea/agora/blob/v0.x.x/source/scpp/README.md) and each file has a `// Copyright 2019 Stellar Development Foundation and contributors. [...]` comment at the top for it. It would be pretty hard to argue we are trying to hide the origin of the code.

But I am completely fine with your approach, just wanted to give context into why I am not concerned with this myself.

**Comment by Geod24** [comment]
There seem to be a POSIX specific mangling bug:
```
ld: warning: directory not found for option '-L@@HOMEBREW_CELLAR@@/libsodium/1.0.18/lib'
Undefined symbols for architecture x86_64:
  "stellar::quorum_tracker_rebuild(stellar::QuorumTracker*, std::__1::shared_ptr<stellar::SCPQuorumSet> (*)(void*, stellar::PublicKey const&), void*)", referenced from:
      __D4scpd6quorum13QuorumTrackerQp7rebuildMFNeDRNbNiNeKxSQCa5types13Stellar_types9PublicKeyZSQDk3Cpp__T10shared_ptrTSQEiQCi11Stellar_SCP12SCPQuorumSetZQBwZv in agora-unittests.o
  "stellar::QuorumIntersectionChecker::create(std::unordered_map<stellar::PublicKey, std::__1::shared_ptr<stellar::SCPQuorumSet>, std::__1::hash<stellar::PublicKey>, std::__1::equal_to<stellar::PublicKey>, std::__1::allocator<std::__1::pair<stellar::PublicKey const, std::__1::shared_ptr<stellar::SCPQuorumSet> > > > const&)", referenced from:
      __D4scpd6quorum23QuorumIntersectionTests17__unittest_L20_C1FZv in agora-unittests.o
      __D4scpd6quorum23QuorumIntersectionTests17__unittest_L39_C1FZv in agora-unittests.o
      __D4scpd6quorum23QuorumIntersectionTests17__unittest_L78_C1FZv in agora-unittests.o
      __D4scpd6quorum23QuorumIntersectionTests18__unittest_L101_C1FZv in agora-unittests.o
      __D4scpd6quorum23QuorumIntersectionTests18__unittest_L143_C1FZv in agora-unittests.o
      __D4scpd6quorum23QuorumIntersectionTests18__unittest_L314_C1FZv in agora-unittests.o
      __D4scpd6quorum23QuorumIntersectionTests18__unittest_L327_C1FZv in agora-unittests.o
      ...
ld: symbol(s) not found for architecture x86_64
```

**Comment by AndrejMitrovic** [comment]
Yeah that is legitimate. I also don't want to increase my stats with fake line counts.

**Comment by Geod24** [comment]
Since it's OSX only, perhaps the `std::shared_ptr` mangling ?

**Comment by Geod24** [comment]
> Yeah that is legitimate. I also don't want to increase my stats with fake line counts.

Ah didn't consider that. Very good point.

**Comment by AndrejMitrovic** [comment]
It would be cool to have those build artifacts so we can `nm` the objects.. since I'm not on a mac right now.

**Comment by TrustHenry** [comment]
Currently, the Windows CI test is marked as green because there is a bug.

Mangling bugs keep popping up in Windows CI Test.  :(

**Comment by AndrejMitrovic** [comment]
I know. I will implement a workaround for the mangling issue.

**Comment by AndrejMitrovic** [comment]
I thought the issue was with the return value which was `shared_ptr`, but actually it might be the `unordered_map` which in this case stores a `shared_ptr` as its value.

I see `__1` in the mangling, maybe that's what's throwing it off?

**Comment by AndrejMitrovic** [comment]
ok I found the linker issue, will submit a separate PR to fix it.

**Comment by AndrejMitrovic** [comment]
Completed fixes after self-review.

**Comment by AndrejMitrovic** [comment]
With regards to `Make Quorum checker buildable`, note that doing this upstream isn't easy because the bindings we're targeting here is for v11.2.0, whereas the one in stellar-core master branch contains even more specialized code which we can't easily bind to. For example, they use interrupts to stop the quorum checker (because sometimes it takes too long to finish).

File: source/scpp/README.md [hunk 16]
```diff
@@ -15,6 +15,8 @@ Some files have been modified after porting:
 
 - Some `private` declarations were changed to `public`. This allows us to test size & layout checks for our D glue layer, for example: https://github.com/bpfkorea/agora/blob/76e4ca7d77b123ef867d3ce99c1c1b26afab761d/source/scpp/src/crypto/ByteSlice.h#L21 used in https://github.com/bpfkorea/agora/blob/76e4ca7d77b123ef867d3ce99c1c1b26afab761d/source/scpp/extra/DSizeChecks.cpp and https://github.com/bpfkorea/agora/blob/76e4ca7d77b123ef867d3ce99c1c1b26afab761d/source/scpp/extra/DLayoutChecks.cpp.
 
+- The `src/quorum` folder contains the Quorum intersection code which is used in stellar-core. It was adapted to not depend on any stellar-specific type definitions. The `update.d` script does not track these files yet, so they have to be kept manually in sync when updating the bindings.
```
Comment (Geod24) [inline source/scpp/README.md:18]: Was it taken from the same commit ?

Comment (AndrejMitrovic) [reply source/scpp/README.md:18]: from 11.2.0

File: source/scpd/quorum/QuorumIntersectionChecker.d [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+nothrow @trusted @nogc:
+
+import scpd.Cpp;
+import scpd.types.Stellar_types;
+import scpd.quorum.QuorumTracker;
+
+extern (C++, `stellar`):
+
+/// Ditto
+extern(C++, class) public abstract class QuorumIntersectionChecker
```
Comment (Geod24) [inline source/scpd/quorum/QuorumIntersectionChecker.d]:
```suggestion
public abstract class QuorumIntersectionChecker
```

File: source/scpd/quorum/QuorumIntersectionChecker.d [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+*******************************************************************************/
+
+module scpd.quorum.QuorumIntersectionChecker;
+
+nothrow @trusted @nogc:
+
+import scpd.Cpp;
+import scpd.types.Stellar_types;
+import scpd.quorum.QuorumTracker;
```
Comment (Geod24) [inline source/scpd/quorum/QuorumIntersectionChecker.d]:
```suggestion
import scpd.quorum.QuorumTracker;
import scpd.types.Stellar_types;
```

File: source/scpd/quorum/QuorumTracker.d [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    /// returns false on failure
+    /// if expand fails, the caller should instead use `rebuild`
+    bool expand(const ref NodeID id, SCPQuorumSetPtr qSet);
+
+    /// Lookup function to find a mapping from Node ID => Quorum set
+    private alias CallbackType = SCPQuorumSetPtr delegate(const ref NodeID) @trusted @nogc nothrow;
+
+    /// returns the known quorum map of the entire network
+    ref const(QuorumMap) getQuorum() const;
+}
```
Comment (Geod24) [inline source/scpd/quorum/QuorumTracker.d:72]:
```suggestion
}

static assert(QuorumTracker.sizeof == 48);
```

Provided this is made a `struct` instead of a `class`. But this won't pass because of #753, which this uncovered.

File: source/scpd/quorum/QuorumTracker.d (hunk 10 above)
Comment (Geod24) [reply source/scpd/quorum/QuorumTracker.d]: This was not addressed

```suggestion
    bool expand (const ref NodeID id, SCPQuorumSetPtr qSet);
```

File: source/scpd/quorum/QuorumTracker.d [hunk 20]
```diff
@@ -0,0 +1,67 @@
+/*******************************************************************************
+
+    Bindings for quorum/QuorumTracker.h
+
+    Note: D code does not need to use anything in this class except the
```
Comment (Geod24) [inline source/scpd/quorum/QuorumTracker.d]:
```suggestion
    D code does not need to use anything in this class except the
```

Having `Note: ` will create a section which will look bad in the doc IIRC. If you really intended this, perhaps
```
Note:
D code does not [...]
```
is more appropriate

File: source/scpd/quorum/QuorumTracker.d [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+extern (C++, `stellar`):
+
+/// helper class to help track the overall quorum over time
+/// a node tracked is definitely in the transitive quorum.
+/// If its associated quorum set is empty (nullptr), it just means
+/// that another node has that node in its quorum set
+/// but could not explore the quorum further (as we're missing the quorum set)
+/// Nodes can be added one by one (calling `expand`, most efficient)
+/// or the quorum can be rebuilt from scratch by using a lookup function
+extern(C++, class) public class QuorumTracker
```
Comment (Geod24) [inline source/scpd/quorum/QuorumTracker.d]:
```suggestion
extern(C++, class) public struct QuorumTracker
```

I really need to write that C++ documentation to explain how to bind types. Usually if a type has no single virtual method, it's better to go with `struct`, unless you know it's only used as a reference type. Even then, if it's passed by `ref` instead of `ptr`, game over.

File: source/scpd/quorum/QuorumTracker.d [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+private:
+    SCP* mSCP;
+    QuorumMap mQuorum;
+
+public:
+    /// Initialize a new QuorumTracker
+    this (SCP* scp);
+
+    /// returns true if id is in transitive quorum for sure
+    bool isNodeDefinitelyInQuorum(const ref NodeID id);
```
Comment (Geod24) [inline source/scpd/quorum/QuorumTracker.d]: This is not correct, and neither are the following ones. In D those functions are virtual, while in C++ they are not. You can make either the affected functions `final`, or, if the class doesn't have a single `virtual` function, make the `class` itself final.

From my experience, things declared as `class` with no `virtual` method => `struct`. And even more so if the `class` is stored as a member of another type.

File: source/scpd/quorum/QuorumTracker.d [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+    /// and add its dependencies as defined by `qset`
+    /// returns true if expansion succeeded
+    ///     `id` was unknown
+    ///     `id` was known and didn't have a quorumset
+    /// returns false on failure
+    /// if expand fails, the caller should instead use `rebuild`
+    bool expand(const ref NodeID id, SCPQuorumSetPtr qSet);
+
+    /// Lookup function to find a mapping from Node ID => Quorum set
+    private alias CallbackType = SCPQuorumSetPtr delegate(const ref NodeID) @trusted @nogc nothrow;
```
Comment (Geod24) [inline source/scpd/quorum/QuorumTracker.d]: This is not used

File: source/scpd/quorum/QuorumIntersectionChecker.d [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+
+    /// Returns: true if the network enjoys quorum intersection
+    abstract bool networkEnjoysQuorumIntersection ();
+
+    /// Returns: the total number of quorums found
+    abstract size_t getMaxQuorumsFound ();
+
+    /// Returns: A pair of possible quorum splits found, or empty pair if none
+    abstract pair!(vector!NodeID, vector!NodeID) getPotentialSplit ();
+}
```
Comment (Geod24) [inline source/scpd/quorum/QuorumIntersectionChecker.d:42]:
```suggestion
}

static assert(__traits(classInstanceSize, QuorumIntersectionChecker) == 8);
```

File: source/scpd/quorum/QuorumTracker.d [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+{
+public:
+    /// The Node ID => Quorum set map type
+    alias QuorumMap = unordered_map!(NodeID, SCPQuorumSetPtr);
+
+private:
+    SCP* mSCP;
+    QuorumMap mQuorum;
+
+public:
```
Comment (Geod24) [inline source/scpd/quorum/QuorumTracker.d:50]: I don't mind the usage of `:`-attributes, but can we merge the two `public:` together ?

File: source/scpd/quorum/QuorumIntersectionTests.d [hunk 26]
```diff
@@ -1,315 +1,298 @@
...
+    License:
+        Licensed under the Apache License, Version 2.0. See the COPYING file at
+        http://www.apache.org/licenses/LICENSE-2.0
+
+*******************************************************************************/
+
+module scpd.quorum.QuorumIntersectionTests;
+
+import scpd.quorum.QuorumIntersectionChecker;
+import scpd.quorum.QuorumTracker;
```
Comment (Geod24) [inline source/scpd/quorum/QuorumIntersectionTests.d:22]: Wrong order

**Review by Geod24 — CHANGES_REQUESTED** [review]
Woah... Sorry it took me so long to review this. So I left a few comments. The most important ones are w.r.t. `class` vs `struct`, and the need for size checks which uncovered #753 .

A few more things:
- I didn't port the SCP tests when I imported the SCP code. The reasoning was that little change would be required to SCP (if any), and none of that will be of substance. Any substantial change should be submitted upstream, where its sanity could be checked. So basically piggybacking on upstream having proper tests. I was a bit surprised when I saw the porting of the tests, and would have not originally asked for it, unless you had major plans to change the code being imported (which I guess you do). But the reasoning I just mentioned was also intended to save us time, and if the work is done, I see no reason not to pull it.
- The last two commit should be merged. I appreciate the effort to make the diff readable (and it was much more pleasant that it could have been), but since `dub` compiles *everything*, there will be a commit that doesn't build. Also, I believe `git` (/github) has greatly improved their rename tracking, so the diff might not be so bad (did you test it?). But atomic commit trumps readability in that case.
- As mentioned above, `dub` compiles everything. You might want to add the tests to the `excludedSourceFiles` so is only compiled on unittests.

Given that we use the fake wrapper you mentioned in #753 this is good to go after the fixes. Just add the commented out size check and a comment explaining why it is not currently correct.

**Comment by AndrejMitrovic** [comment]
> I didn't port the SCP tests when I imported the SCP code. 

I wanted to verify the binding was correct. It wasn't a big effort tho, and it did catch bugs in my binding.

That being said I guess we don't really need it anymore. We will have separate tests for the quorums we generate, so maybe we can remove this code?

It might even be complicated to have something which is Stellar-licensed in our codebase, but I'm not lawyer.

> but since dub compiles everything, there will be a commit that doesn't build.

Yeah this was purely to make it easy to review. I'll squash them.

 > As mentioned above, dub compiles everything. You might want to add the tests to the excludedSourceFiles so is only compiled on unittests.

Oh right. Will fix.

File: source/scpd/quorum/QuorumTracker.d (hunk 22 above)
Comment (AndrejMitrovic) [reply source/scpd/quorum/QuorumTracker.d]: > In D those functions are virtual

Oh damn. How come it worked?

> From my experience, things declared as class with no virtual method => struct.

What about classes which inherit other classes? This one specifically is defined as `class QuorumTracker : public NonMovableOrCopyable`, so it looks like it was never supposed to be used as a POD type.

Comment (Geod24) [reply source/scpd/quorum/QuorumTracker.d]: > class QuorumTracker : public NonMovableOrCopyable

Sounds like the idiomatic way to do this in D would be:
```D
mixin template NonMovableOrCopyable (T /* = typeof(this) */) // Not sure if the default arg works
{
    @disable this(this);
    @disable ref T opAssign () (auto ref T other);
}

public struct QuorumTracker
{
    mixin NonMovableOrCopyable!(QuorumTracker);
}
```

That might be a bit of an overreaching statement, but my experience is that good, principled C++ code ends up being idiomatic D code, and trying to use the same construct in D as one does in C++ might lead one astray.

> Oh damn. How come it worked?

Depends on the usage. The code called from C++ doesn't have an issue, obviously. The code called from D may or may not crash, depending on whether or not the compiler devirtualize the call. And since we dropped DMD, there is a good chance it does :)

Comment (Geod24) [reply source/scpd/quorum/QuorumTracker.d]: Note, the above snippet doesn't fully capture the C++ intent though, since in D, you *cannot* declare a `struct` are non movable. Should probably include a copy constructor too but it currently triggers a cascade of bugs.

Comment (AndrejMitrovic) [reply source/scpd/quorum/QuorumTracker.d]: > And since we dropped DMD, there is a good chance it does :)

LOL.

Yeah also I dislike virtual by default in D. Where is Manu when you need him.

File: source/scpd/quorum/QuorumIntersectionChecker.d (hunk 24 above)
Comment (AndrejMitrovic) [reply source/scpd/quorum/QuorumIntersectionChecker.d:42]: Sure, but why?

File: source/scpd/quorum/QuorumTracker.d (hunk 25 above)
Comment (AndrejMitrovic) [reply source/scpd/quorum/QuorumTracker.d:50]: Actually I dislike `:` attributes personally. I'll fix up the mistake though.

**Comment by Geod24** [comment]
> It might even be complicated to have something which is Stellar-licensed in our codebase, but I'm not lawyer.

We already do (SCP), and it's Apache licensed (and we have LICENSE-APACHE.txt included and haven't removed any authorship / license information), so it's all good.

File: source/scpd/quorum/QuorumIntersectionChecker.d (hunk 24 above)
Comment (Geod24) [reply source/scpd/quorum/QuorumIntersectionChecker.d:42]: Why it is `8` or why we should have this check ?

**Comment by Geod24** [comment]
```
agora.obj : error LNK2001: unresolved external symbol "public: virtual __cdecl stellar::QuorumIntersectionChecker::~QuorumIntersectionChecker(void)" (??1QuorumIntersectionChecker@stellar@@UEAA@XZ)
```

**Comment by AndrejMitrovic** [comment]
Also the block storage checksum thing failed on Windows again.

Like I said in https://github.com/bpfkorea/agora/pull/731, why not just version it out on Windows and fix it later?

**Comment by AndrejMitrovic** [comment]
Can I just get rid of the destructor? I don't know why it's defined on the C++ side:

```C++
virtual ~QuorumIntersectionChecker(){};
```

File: source/scpd/quorum/QuorumIntersectionChecker.d (hunk 24 above)
Comment (Geod24) [reply source/scpd/quorum/QuorumIntersectionChecker.d:42]: @AndrejMitrovic answered directly: the later.

Because it's a rather simple check that can save you a lot of headache. For example, out of the two types introduced in this PR, one had the wrong size (`QuorumTracker`) and adding the check uncovered that and #753 . Also makes sure that in the future, we don't break this by accident (e.g. change in alignment, or adding a field in a nested type).

In my experience, those little checks have an immediate benefit.

**Comment by Geod24** [comment]
It's defined so that there is an entry in the vtable for the destructor. Otherwise you end up with things like [this](https://issues.dlang.org/show_bug.cgi?id=18616).
Except in C++ it is `inline` for the base class. Just provide an empty destructor in the D binding and you should be 👍

**Comment by AndrejMitrovic** [comment]
I don't understand where the submodule update is coming from..

**Comment by AndrejMitrovic** [comment]
Ok I managed to get rid of it.

**Review by Geod24 — APPROVED** [review]
Thanks
